File: payharness/packages/sdk-php/tests/WebhookTest.php
<?php

declare(strict_types=1);

use PayHarness\Webhook;

spl_autoload_register(static function (string $class): void {
    $prefix = 'PayHarness\\';
    if (!str_starts_with($class, $prefix)) {
        return;
    }
    $path = dirname(__DIR__) . '/src/' . str_replace('\\', '/', substr($class, strlen($prefix))) . '.php';
    require $path;
});

$secret = 'whsec_test';
$body = '{"type":"payment.succeeded"}';
$timestamp = 1_700_000_000;
$signature = Webhook::sign($secret, $timestamp, $body);

if (!str_starts_with($signature, 't=')) {
    throw new RuntimeException('Webhook signature format is invalid.');
}

if (!Webhook::verify($secret, $signature, $body, $timestamp)) {
    throw new RuntimeException('Valid webhook signature was rejected.');
}

if (Webhook::verify($secret, $signature, '{"type":"payment.failed"}', $timestamp)) {
    throw new RuntimeException('Modified webhook body was accepted.');
}

if (Webhook::verify($secret, $signature, $body, $timestamp + 301)) {
    throw new RuntimeException('Stale webhook signature was accepted.');
}

if (Webhook::verify($secret, 'invalid', $body, $timestamp)) {
    throw new RuntimeException('Malformed webhook signature was accepted.');
}

echo "Webhook tests passed.\n";

File: payharness/packages/sdk-go/webhook_test.go
package payharness

import (
	"testing"
	"time"
)

func TestWebhookSignature(t *testing.T) {
	secret := "whsec_test"
	body := `{"type":"payment.succeeded"}`
	timestamp := int64(1700000000)
	signature := CreateWebhookSignature(secret, timestamp, body)
	if signature[:2] != "t=" {
		t.Fatalf("unexpected signature: %s", signature)
	}
	now := time.Unix(timestamp, 0)
	if !VerifyWebhookSignature(secret, signature, body, now) {
		t.Fatal("expected valid signature")
	}
	if VerifyWebhookSignature(secret, signature, `{"type":"payment.failed"}`, now) {
		t.Fatal("expected modified body to fail")
	}
	if VerifyWebhookSignature(secret, signature, body, now.Add(301*time.Second)) {
		t.Fatal("expected stale signature to fail")
	}
	if VerifyWebhookSignature(secret, "invalid", body, now) {
		t.Fatal("expected malformed signature to fail")
	}
}
